Continuous Integration populated

PeterTurcan opened the pull request:
@anarthal This version should include all the topics you listed in your earlier notes. Not sure I have all the tools (compilers, sanitizers) listed that perhaps should be. And a link to fuzz testing would be good too.
@alandefreitas FYI

alandefreitas commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
I think Sam wouldn't agree with that. Although Ruben would be more favorable to docker in CI. I think we need to list the cases where docker is useful.

Reducing CI times is questionable in most simpler libraries. The workflow usually starts with an image and then downloading a few binaries or apt-get or something. Let's say you need GCC11 as a workflow. 

- As apt-get already gives you compiled GCC11 binaries and any binaries in the image also need to be downloaded (no free lunch), you might end up saving nothing. 
- Now consider the case where the github image already contains GCC11. Now you're actually losing a free-lunch, because you have an image with GCC11 and you are using it to download another image with GCC11.
- Then there's the problem of testing multiple compilers. If you have one docker container for each compiler, then this has a high cost. If you have one docker container with all compilers, now you have all workflows downloading all compilers and then just testing one of them, while the github image would already have most of these compilers.

Consistency and isolation are good features of docker more than reducing CI times, but it also has problems. It goes somewhat against the purpose of one of our missions which is to support the most common environments instead of an isolated environment. The scenario where this is useful is more specific and not the general case and that has to be clear in the text. There are basically two scenarios where isolation is very useful. 

- The first is when we are replicating environments we couldn't replicate otherwise. For instance, we need ubuntu 16 all the time but there's no github image for it. So we use the ubuntu 22 image and run a ubuntu 16 docker container on it. This is the most common case by far.
- The second case where it's very useful is when the workflow depends on tools that can't be easily downloaded with apt-get. One example is mrdox, where we need a very specific version of LLVM. The binaries for this version are not available anywhere and building it from source takes hours. So we have a docker container that bundles the LLVM binaries and everything is much easier. But this use case is also a little questionable because we could also store these binaries elsewhere and download them in CI, which would have the same effect. The best use case for docker seems to be really just replicating environments we couldn't replicate otherwise.

alandefreitas commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
That's a very weird sentence.

PeterTurcan replied:
reworded

alandefreitas reviewed (commented):
Have a look at some content you could reuse from

https://github.com/boostorg/boost-ci/tree/master

It's the "source of truth" boost developers currently have for what they should do in CI.

It's also a typical model of what a library should look like. Developers can start a new functional boost library by just downloading the contents of this repository and adjusting it.

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -16,7 +16,7 @@ Boost has been incorporating GitHub Actions into its testing workflows. This is
-Boost also uses GitHub Actions support for matrix builds (allowing Boost to run the same build steps on multiple combinations of operating systems, compilers, etc.), caching of dependencies to speed up builds, and the ability to create custom actions.
+Boost also uses GitHub Actions support for _matrix builds_ (allowing Boost to run the same build steps on multiple combinations of operating systems, compilers, etc.), caching of dependencies to speed up builds, and the ability to create custom actions.
 
 === Example GitHub Workflows
```
OK, removed the super-project examples, added URL library

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -144,6 +158,165 @@ Refer to:
+
+All the tools listed have different strengths and are useful in different scenarios, so you might want to use several of them in combination.
+
+=== Valgrind
```
Moved Valgrind to last.

PeterTurcan commented:
@alandefreitas I have simplified the CI section:
1. moved static analysis to a separate topic (except for Coverity Scan as its embedded in our CI)
2. removed the CI perf stuff, didn't seem to be helpful
3. added a link and short explanation to boost-ci, rather than duplicate its content
4. added summaries of tools used in boost-ci (coverity scan, codecov.io) that were missing from the topic
5. simplified the Docker use case to the replication issue only

alandefreitas commented:
Great. Anything left to review?

fpelliccioni commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 = Continuous Integration
 
+
 The Boost project uses Continuous Integration (CI) to ensure the quality and integrity of its code. CI is the practice of merging all developers' working copies to a shared mainline several times a day. The main aim is to prevent integration problems, which are detected and can be fixed as early as possible.
```
```suggestion
The Boost project utilizes Continuous Integration/Continuous Deployment (CI/CD) to ensure the quality and integrity of its code. CI/CD is the practice of merging all developers' working copies into a shared mainline several times a day. The primary objective is to prevent integration issues, which can be identified and addressed as soon as possible.
```

fpelliccioni commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 
 Boost uses several CI services for testing on different platforms and compilers. Many libraries use two or three of the systems described here, as does the https://github.com/boostorg/boost/tree/master[Super-project] itself.
 
 Note:: It is a requirement for a new library submission to Boost to include an appropriate CI system. Refer to the examples for each CI system to better understand what is involved.
```
```suggestion
Note:: It is a requirement for a new library submission to Boost to include an appropriate CI/CD system. Refer to the examples for each CI/CD system to better understand what is involved.
```

PeterTurcan replied:
I am leaving it as CI for now - not sure we use the CD side of these tools as the releases are not updated? Open to correction.

fpelliccioni commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 == GitHub Actions
 
-Boost has been incorporating GitHub Actions into its testing workflows. This is a Continuous Integration/Continuous Deployment (CI/CD) system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
+Boost has been incorporating https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions[GitHub Actions] into its testing workflows. This is a Continuous Integration/Continuous Deployment (CI/CD) system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
```
```suggestion
Boost has been incorporating https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions[GitHub Actions] into its testing workflows. This is a CI/CD system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
```

PeterTurcan replied:
done

fpelliccioni commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -130,20 +142,175 @@ CircleCI also supports parallel testing, caching of dependencies, and matrix bui
+
+. Azure Pipelines uses a file named `.azure-pipelines.yml` at the root of the repository to define the build environment and steps. This file specifies the operating systems and compilers to use, any dependencies to install, and the commands to run for building and testing the code.
+
+. When changes are pushed to the repository, Azure Pipelines automatically initiates a build and runs the tests according to the instructions in `azure-pipelines.yml`. This helps ensure that changes do not break existing functionality.
```
```suggestion
. When changes are pushed to the repository, Azure Pipelines automatically triggers a build and runs the tests according to the instructions in the `.azure-pipelines.yml` file. This helps ensure that changes do not break existing functionality.
```

PeterTurcan replied:
done

fpelliccioni commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 = Continuous Integration
```
Cirrus CI (https://cirrus-ci.org/) offers FreeBSD support. Considering its unique capability, could it be worth incorporating into the docs and the Boost continuous integration process?

PeterTurcan replied:
@fpelliccioni Probably a good question for @alandefreitas  or @anarthal ?

alandefreitas replied:
Nothing against that. I just haven't seen it being used in other libraries yet.

anarthal commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -130,20 +142,175 @@ CircleCI also supports parallel testing, caching of dependencies, and matrix bui
 * https://github.com/boostorg/multiprecision/blob/380aae3c28c646ea2ca1b42156d83732295082d7/.circleci/config.yml[Multiprecision library config.yml]
 
-=== CircleCI Reference
+== Codecov.io
```
I think coverage deserves its own section, and that we should expand on how to set it up correctly.

PeterTurcan replied:
Created a coverage section, with a new intro.

anarthal commented on contributor-guide/modules/ROOT/pages/testing/static-analysis.adoc:
```diff
@@ -0,0 +1,56 @@
+
+== Treat Warnings as Errors
+
+One of the most important things for library developers to do is to turn on compiler settings to treat Warnings as Errors. In practice, some of these warnings will need to be whitelisted as they will be false positives. Otherwise, most warnings should be addressed as errors.
```
Please mention the `warnings-as-errors=on` feature in B2. While there is a [`CMAKE_COMPILE_WARNING_AS_ERROR`](https://cmake.org/cmake/help/latest/variable/CMAKE_COMPILE_WARNING_AS_ERROR.html#variable:CMAKE_COMPILE_WARNING_AS_ERROR), it's 3.24 >= so may not be an option

PeterTurcan replied:
ok

anarthal commented on contributor-guide/modules/ROOT/pages/testing/static-analysis.adoc:
```diff
@@ -0,0 +1,56 @@
+
+One of the most important things for library developers to do is to turn on compiler settings to treat Warnings as Errors. In practice, some of these warnings will need to be whitelisted as they will be false positives. Otherwise, most warnings should be addressed as errors.
+
+== Stand-alone Static Analysis Tools
```
Which libraries use any of the below tools? Please link

PeterTurcan replied:
@anarthal Where would I find out this information, is it in the boostorg repo somewhere?

alandefreitas replied:
I use a lot of these tools locally but have never seen them in CI. I don't even know of a good way to integrate static analysis in CI, to be honest.

anarthal replied:
I don't have an answer either - I'd be the first one interested in adding such things. I suspect there aren't libs using them yet. I'd just not mention them if there are no examples - we can add them later if we start using them for any lib.

alandefreitas replied:
> I'd be the first one interested in adding such things

I wouldn't. I love static analysis locally, but I ignore most of the errors because these tools (correctly) allow themselves to have false positives and warn about recommendations rather than real errors. This is very useful locally because I just go disabling the warnings that don't help me. But that would be chaos in CI. Nothing would pass.

anarthal commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 == GitHub Actions
 
-Boost has been incorporating GitHub Actions into its testing workflows. This is a Continuous Integration/Continuous Deployment (CI/CD) system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
+Boost has been incorporating https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions[GitHub Actions] into its testing workflows. This is a Continuous Integration/Continuous Deployment (CI/CD) system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
```
I'd note the main problem we've got with GHA nowadays - since all builds run under boostorg, we hit the quota and may take a long time

PeterTurcan replied:
note added

anarthal commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
I think this needs rework. I'd move this to a separate document.

* You're mentioning clang's ASAN but linking to MSVC's ASAN. I'd just mention clang's. I don't know if we support building with ASAN's MSVC via any of our tools.
* Saying that the four sanitizers are part of the address sanitizer is a little bit confusing. They're just different sanitizers.
* ASAN is enabled by passing the `-fsanitize=address` to clang, or with the `<address-sanitizer>on` B2 flag. Under cmake, you need to add the flags manually. I'd mention leak sanitizer together here, since it runs with asan. The standalone version is for advanced scenarios I'd say, I don't think it's worth mentioning it here. https://clang.llvm.org/docs/AddressSanitizer.html
* I think Boost.CI doesn't include any build for msan does it? I think it's good mentioning here, too. B2 doesn't have a feature for it, so you need to specify `-fsanitize=memory`. https://clang.llvm.org/docs/MemorySanitizer.html
* UBSan `-fsanitize=undefined` or `undefined-sanitizer=on` in B2. https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
* TSan `-fsanitize=thread` or `thread-sanitizer=on` in B2. https://clang.llvm.org/docs/ThreadSanitizer.html

PeterTurcan replied:
OK, moved sanitizers to a separate page.

PeterTurcan replied:
And added your notes for each santizer.

anarthal commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 
 Note:: It is a requirement for a new library submission to Boost to include an appropriate CI system. Refer to the examples for each CI system to better understand what is involved.
 
+== New Library CI Framework
```
I think this is missing the Boost-specific details, which are what makes CI testing in Boost difficult in my opinion. Some things that were really hard for me:
* You test with the superproject, cloning it and placing your lib inside. No-one else does this, you normally would consume Boost as a dependency.
* You test with the develop or master branches of the superproject, instead of a stable release.
* The cmake workflows are not trivial and not usual cmake.

I don't think we can expect @PeterTurcan to know these things because they are very Boost specific, as I recall Vinnie said. This is a really nice start for the CI documentation, but I think me, @alandefreitas or any other author should add the nitty-gritty details this is missing.

PeterTurcan replied:
These seem like valid points to make, so added them. More detail from lib authors can be added in due course.

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -104,21 +122,15 @@ AppVeyor also supports parallel testing, a build cache to speed up builds, and t
 == CircleCI
 
-CircleCI is a CI/CD platform that supports a wide range of languages, tools, and services, making it flexible for different testing requirements. It is less commonly used than <<Travis CI>> or <<AppVeyor>>, but is used by the Super-project and a few libraries.
+https://circleci.com/developer[CircleCI] is a CI/CD platform that supports a wide range of languages, tools, and services, making it flexible for different testing requirements. It is less commonly used than <<Travis CI>> or <<AppVeyor>>, but is used by the Super-project and a few libraries.
```
ok

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 == Travis CI
 
-Travis CI is used for testing on Linux and macOS environments. It is a hosted, distributed continuous integration service used to build and test software projects hosted at GitHub. Here's the overall process:
+https://docs.travis-ci.com/user/for-beginners/[Travis CI] is used for testing on Linux and macOS environments. It is a hosted, distributed continuous integration service used to build and test software projects hosted at GitHub. Here's the overall process:
```
Added a note it is not recommended for new libs.

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
-. Drone uses a file named `.drone.jsonnet` or `.drone.star`, and a `.drone` folder, at the root of the repository to define the build pipeline, including the environment, build steps, and notification settings. The environment is typically a Docker container, allowing for a high degree of flexibility and customization.
+* https://github.com/bazelbuild/starlark[Starlark] is a Python-inspired language that was created by Google and is used for configuring Bazel build systems. Similar to Jsonnet, it allows you to create more complex and maintainable pipeline configurations.
+
+Both Jsonnet and Starlark allow you to use logic like conditionals and loops in your configuration files, making them a powerful tool for complex CI/CD pipelines. If your team has Python experience, for example, Starlark's Python-like syntax might be a better choice. If you're working in a JSON-heavy environment, Jsonnet might be more suitable.
```
changed to singular

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
 == Drone
 
-Drone is an open-source CI system built on container technology. Each step in a Drone build usually runs in its own Docker container, allowing it to use any language, tool, or service that can run inside a Docker container. This offers excellent environment consistency and isolation between steps. You can run your Drone pipelines locally on your machine for testing purposes before pushing changes ...
+https://docs.drone.io/[Drone] is an open-source CI system built on container technology. Each step in a Drone build usually runs in its own Docker container, allowing it to use any language, tool, or service that can run inside a Docker container. This offers excellent environment consistency and isolation between steps. You can run your Drone pipelines locally on your machine for testing purpose ...
```
Added a note, and links

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -104,21 +122,15 @@ AppVeyor also supports parallel testing, a build cache to speed up builds, and t
 == CircleCI
 
-CircleCI is a CI/CD platform that supports a wide range of languages, tools, and services, making it flexible for different testing requirements. It is less commonly used than <<Travis CI>> or <<AppVeyor>>, but is used by the Super-project and a few libraries.
+https://circleci.com/developer[CircleCI] is a CI/CD platform that supports a wide range of languages, tools, and services, making it flexible for different testing requirements. It is less commonly used than <<Travis CI>> or <<AppVeyor>>, but is used by the Super-project and a few libraries.
```
linked

PeterTurcan commented on contributor-guide/modules/ROOT/pages/testing/continuous-integration.adoc:
```diff
@@ -1,71 +1,94 @@
-Boost has been incorporating GitHub Actions into its testing workflows. This is a Continuous Integration/Continuous Deployment (CI/CD) system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
+Boost has been incorporating https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions[GitHub Actions] into its testing workflows. This is a Continuous Integration/Continuous Deployment (CI/CD) system platform provided directly by GitHub. It can run tests on a variety of platforms and configurations. Here's a basic outline of how GitHub Actions works for Boost:
 
 . GitHub Actions use YAML files stored in a directory called `.github/workflows/` at the root of the repository to define the build environment and steps. For instance, a workflow file might specify which operating systems and compilers to use, any dependencies to install, and the commands to run for building and testing the code.
```
done

PeterTurcan commented:
@alandefreitas @anarthal @fpelliccioni 
CI, sanitizers, static analysis separated. Your comments included.

anarthal commented:
Looks good

PeterTurcan commented:
@alandefreitas I have removed the static analysis section for now. Squashed the remainder, so I hope GTG.
